### DISCUSSION TITLE
Postgres with Trusted Lang Extensions (TLE)

### DISCUSSION THREAD

**Discussion by bobmcwhirter** [guacsec/trustify#94] [Ideas]
I think it could be useful to evaluate TLE from Amazon.

https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/PostgreSQL_trusted_language_extension.html

While this is RDS-centric docs, it is possible to run the same in a local Pgsql.

We maybe should spin a ... strategy? for our embedded-pgsql to use a pgsql+tle so that we could map directly to TLE in RDS for deployments targeting AWS.

This would all us to freely write stored procedures and functions inside of postgresql using Rust.

**Comment by helio-frota** [comment]
@bobmcwhirter this example is working: https://github.com/helio-frota/plrust-example

just clone + copy/pasta commands to evaluate :+1:

**Reply by bobmcwhirter** [reply]
Hey, excellent!

Do we imagine we can crate-ify this to work with embedded-postgresql?

**Reply by helio-frota** [reply]
tbqh I don't know :man_shrugging:  (it contains some weirdo points)

1) The user `postgres` it is required to build/compile/install https://github.com/helio-frota/plrust-example/blob/main/Containerfile#L18
1.1) we can't compile the plrust-function using our rust versions inside IDE for example
1.2) using plrust also means using pgrx and it contains [this list](https://github.com/pgcentralfoundation/pgrx/tree/master?tab=readme-ov-file#caveats--known-issues)
2) Required to use rust [1.72.0](https://github.com/helio-frota/plrust-example/blob/main/Containerfile#L20) at the moment ( also in [online docs](https://plrust.io/install-prerequisites.html#clone-plrust-and-check-rust-version) ) to build the plrust.
3) They are using PG15 in their Dockerfile https://github.com/tcdi/plrust/blob/main/Dockerfile.try
3.1) The example I shared is a simplified version (of this ^ Dockerfile) to evaluate with PG16

I'm just sharing things to take a look , not to make influence as I really don't know 

In a simple form they have `extensions` and `functions` 

Trying to answer your question: 

I think that is more possible to have an `extension` (pgrx)  to work with embedded-postgresql than a `function` (plrust)

example https://github.com/VADOSWARE/pg_idkit?tab=readme-ov-file#installing-pg_idkit

Binary install section ^ -- And we need to modify `postgresql.conf` from embedded-postgresql

**Reply by helio-frota** [reply]
this is other extension, as example,  https://github.com/rustprooflabs/pgfaker/tree/main?tab=readme-ov-file#getting-started
to generate fake data. Although release artifact is a .deb file

**Reply by helio-frota** [reply]
other thing to keep in mind is when changing the config file , pg needs restart
https://plrust.io/install-plrust.html?highlight=restart#configure-and-restart-postgresql

and that is required https://github.com/helio-frota/plrust-example/blob/main/init.sh
( in the case of container example the file was changed first ^ and pg started later )

that means we need a modified version of https://github.com/theseus-rs/postgresql-embedded to avoid starting 2x during `cargo run trustd`

**Reply by helio-frota** [reply]
Basic examples added :+1: 

https://github.com/helio-frota/plrust-example?tab=readme-ov-file#basic-examples

The first function takes 3+ minutes to compile

**Comment by bobmcwhirter** [comment]
What I mentioned on today's call

https://github.com/theseus-rs/postgresql-embedded/issues/47#issuecomment-2179792142

**Comment by helio-frota** [comment]
closing, thanks for this task :+1: :
